src/python/pants/core/util_rules/gen_rule.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import shlex
+from typing import cast
+
+from pants.core.target_types import GenRuleCommandField, GenRuleTarget
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.process import (
+    BinaryNotFoundError,
+    BinaryPathRequest,
+    BinaryPaths,
+    Process,
+    ProcessResult,
+    SearchPath,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import Sources
+from pants.util.logging import LogLevel
+
+
+@rule(desc="Running gen_rule", level=LogLevel.DEBUG)
+async def run_gen_rule(gen_rule: GenRuleTarget) -> ProcessResult:
+    args = shlex.split(cast(str, gen_rule[GenRuleCommandField].value))
+    binary_request = BinaryPathRequest(
+        binary_name=args[0],
+        search_path=SearchPath(("/usr/bin", "/bin", "/usr/local/bin")),
+    )
```

Comment:
Interesting! I didn't expect this to open with tool discovery, but I suppose that that might make sense?

One thing that occurs to me though is that it is fairly likely to need multiple tools available to a shell script (for example, if arg0 is `bash`, then you are likely to also need `grep`, `cat` etc). Perhaps rather than automatically discovering `args[0]`, the target type should instead accept a list of tool names to resolve, and then expose them as environment variables that the command can use?